samples/Common.c

```diff
@@ -704,24 +715,70 @@ VOID sampleBandwidthEstimationHandler(UINT64 customData, DOUBLE maximumBitrate)
-        // decrease encoder bitrate by packet loss percent
-        bitrate *= (1.0f - percentLost / 100.0f);
-    }
-    // otherwise keep bitrate the same
-
-    DLOGV("received sender bitrate estimation: suggested bitrate %u sent: %u bytes %u packets received: %u bytes %u packets in %lu msec", bitrate,
-          txBytes, txPacketsCnt, rxBytes, rxPacketsCnt, duration / 10000ULL);
+    UINT8 percentLost = (txPacketsCnt > 0) ? (lostPacketsCnt * 100 / txPacketsCnt) : 0;
+    SampleStreamingSession* pSampleStreamingSession = (SampleStreamingSession*) customData;
+    if (pSampleStreamingSession == NULL) {
+        DLOGW("Invalid streaming session (NULL object)");
+        return;
+    }
+
+    // Calculate packet loss
+    DOUBLE currentPacketLoss = (DOUBLE) percentLost;
+    pSampleStreamingSession->twccMetadata.averagePacketLoss =
+        EMA_ACCUMULATOR_GET_NEXT(pSampleStreamingSession->twccMetadata.averagePacketLoss, currentPacketLoss);
+
+    UINT64 currentTimeMs = GETTIME();
+    UINT64 timeDiff = currentTimeMs - pSampleStreamingSession->twccMetadata.lastAdjustmentTimeMs;
+    if (timeDiff < ADJUSTMENT_INTERVAL_SECONDS) {
+        // Too soon for another adjustment
+        return;
+    }
+
+    MUTEX_LOCK(pSampleStreamingSession->twccMetadata.updateLock);
+    locked = TRUE;
+    videoBitrate = pSampleStreamingSession->twccMetadata.currentVideoBitrate;
+    audioBitrate = pSampleStreamingSession->twccMetadata.currentAudioBitrate;
+    if (locked) {
+        MUTEX_UNLOCK(pSampleStreamingSession->twccMetadata.updateLock);
+        locked = FALSE;
+    }
+
+    if (pSampleStreamingSession->twccMetadata.averagePacketLoss <= 5) {
+        // increase encoder bitrate by 5 percent with a cap at MAX_BITRATE
+        videoBitrate = (UINT64) MIN(videoBitrate * 1.05f, MAX_VIDEO_BITRATE_KBPS);
+    } else {
+        // decrease encoder bitrate by average packet loss percent, with a cap at MIN_BITRATE
+        videoBitrate = (UINT64) MAX(videoBitrate * (1.0f - pSampleStreamingSession->twccMetadata.averagePacketLoss / 100.0f), MIN_VIDEO_BITRATE_KBPS);
+    }
+
+    if (pSampleStreamingSession->twccMetadata.averagePacketLoss <= 5) {
+        // increase encoder bitrate by 5 percent with a cap at MAX_BITRATE
+        audioBitrate = (UINT64) MIN(audioBitrate * 1.05f, MAX_AUDIO_BITRATE_BPS);
+    } else {
+        // decrease encoder bitrate by average packet loss percent, with a cap at MIN_BITRATE
+        audioBitrate = (UINT64) MAX(audioBitrate * (1.0f - pSampleStreamingSession->twccMetadata.averagePacketLoss / 100.0f), MIN_AUDIO_BITRATE_BPS);
+    }
+
+    MUTEX_LOCK(pSampleStreamingSession->twccMetadata.updateLock);
+    locked = TRUE;
+    // Update the session with the new bitrate and adjustment time
+    pSampleStreamingSession->twccMetadata.newVideoBitrate = videoBitrate;
+    pSampleStreamingSession->twccMetadata.newAudioBitrate = audioBitrate;
+    if (locked) {
+        MUTEX_UNLOCK(pSampleStreamingSession->twccMetadata.updateLock);
```

Comment:
We are doing lock/unlock multiple times and only doing assignments in between, there is not much computations. Single lock and unlock should suffice.

Reply:
You are right there is no computation and just assignment. But, this lock is being used in the gstreamer callback while sending frame, so just did not want to block it while updating. Hence the locking only while updating critical members.

Reply:
It is locked twice in a function. And between two locks there is no other function call and no time consuming computations. You can save some CPU cycles by locking once. Lock at beginning and release in the end.